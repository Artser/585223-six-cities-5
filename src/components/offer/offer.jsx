import React, {PureComponent} from "react";
import ReviewForm from "../review-form/review-form";
import ReviewList from "../review-list/review-list";
import Coord from "../coord/coord";
import PropTypes from "prop-types";
import withReviewForm from "../../hocs/with-review-form";
import {Operation} from "../../reducer/data";
import {connect} from "react-redux";
import {offerType} from "../../types";
import {Header} from "../header/header";
// import {getCurrentOffer} from '../../reducer/reselect';
import NearPlace from "../../components/near-places/near-places";
import {FavoriteStatus} from "../../utils/functions";
const ReviewFormWrapped = withReviewForm(ReviewForm);
import {getRating} from '../../utils/functions';


class Offer extends PureComponent {
  constructor(props) {

    super(props);
    this.handleButton = this.handleButton.bind(this);
  }

  handleButton(evt) {
    evt.preventDefault();
    const {updateFavorite, offer} = this.props;
    updateFavorite(offer.id, offer.isFavorite ? FavoriteStatus.MINUS : FavoriteStatus.PLUS);
  }

  componentDidMount() {

    const id = this.props.match.params.id;
    this.props.loadCurrentOffer(id);
    this.props.loadNearPlaces(id);
  }


  render() {
    const {offer, nearPlaces} = this.props;
    if (offer === null) {
      return <p >Loading</p>;

    }
    const bookmarkClass = offer.isFavorite ? `property__bookmark-button--active` : `property__bookmark-button`;
    const ImageBlock = ({img}) => {
      return (
        <div className="property__image-wrapper">
          <img className="property__image" src={img} alt="Photo studio" />
        </div>
      );
    };
    const Goods = ({good}) => {
      return (
        <li className="property__inside-item">
          {good}
        </li>
      );
    };
    return (
      <div className="page">
        <Header
          authorizationStatus={this.props.authorizationStatus}
          authInfo={this.props.authInfo}
        />

        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                {
                  offer.images.map((img, key) => (
                    <ImageBlock
                      key={key}
                      img={img}
                    />
                  ))
                }

              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                <div className="property__mark">
                  <span>{offer.isPremium}</span>
                </div>
                <div className="property__name-wrapper">
                  <h1 className="property__name">
                    {offer.title}
                  </h1>
                  <button onClick={this.handleButton} className={`${bookmarkClass} button`} type="button">
                    <svg className="property__bookmark-icon" width="31" height="33">
                      <use xlinkHref="#icon-bookmark"></use>
                    </svg>
                    <span className="visually-hidden">To bookmarks</span>
                  </button>
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: getRating(offer.rating)}}></span>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value"> {offer.rating}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">
                    {offer.type}
                  </li>
                  <li className="property__feature property__feature--bedrooms">
                    {offer.bedrooms} Bedrooms
                  </li>
                  <li className="property__feature property__feature--adults">
                    Max {offer.adults} adults
                  </li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{offer.price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <ul className="property__inside-list">

                    {
                      offer.goods.map((goods, key) => (
                        <Goods
                          key={key}
                          good={goods}
                        />

                      ))
                    }

                  </ul>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={offer.avatar} width="74" height="74" alt="Host avatar" />
                    </div>
                    <span className="property__user-name">
                      {offer.name}
                    </span>
                  </div>
                  <div className="property__description">
                    <p className="property__text">
                      {offer.description}
                    </p>

                  </div>
                </div>
                <section className="property__reviews reviews">
                  <ReviewList
                  />

                  <ReviewFormWrapped
                    id={this.props.match.params.id}
                  />

                </section>
              </div>
            </div>
            <div style={{height: `579px`}}>
              <Coord


                coords={nearPlaces.map((nearPlace) => (nearPlace.coord))}


                activeOffer={offer.coord}

              />
            </div>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <div className="near-places__list places__list">
                {
                  nearPlaces.map((nearPlace) => (
                    <NearPlace
                      key={nearPlace.id}
                      type={nearPlace.type}
                      title={nearPlace.title}
                      price={nearPlace.price}
                      imgLink={nearPlace.imgLink}
                      rating={nearPlace.rating}
                    />
                  ))
                }

              </div>
            </section>
          </div>
        </main>
      </div>

    );
  }
}
Offer.propTypes = {
  match: PropTypes.shape({params: PropTypes.shape({id: PropTypes.string})}),
  offer: offerType,
  updateFavorite: PropTypes.func,
  img: PropTypes.object,
  good: PropTypes.string,
  loadCurrentOffer: PropTypes.func,
  loadNearPlaces: PropTypes.func,
  authorizationStatus: PropTypes.string,
  authInfo: PropTypes.object,
  nearPlaces: PropTypes.array
};
const mapStateToProps = (state) => {
  return {
    offer: state.activeOffer,
    authorizationStatus: state.authorizationStatus,
    authInfo: state.authInfo,
    nearPlaces: state.nearPlaces,
  };
};

const mapDispatchToProps = (dispatch) => ({
  updateFavorite: (offerId, status) => {
    dispatch(Operation.postFavorite(offerId, status));
  },
  loadNearPlaces: (id) => {
    dispatch(Operation.loadNearPlacesId(id));
  },

  loadCurrentOffer: (id) => {

    dispatch(Operation.loadOfferById(id));
  }
});
export default connect(mapStateToProps, mapDispatchToProps)(Offer);
