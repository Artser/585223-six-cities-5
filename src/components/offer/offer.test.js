import React from 'react';
import renderer from 'react-test-renderer';
import {Offer} from './offer';
import {MemoryRouter} from 'react-router-dom';

jest.mock(`../review-list/review-list`, () => `ReviewList`);
jest.mock(`../review-form/review-form`, () => `ReviewForm`);
jest.mock(`../coord/coord`, () => `Coord`);

describe(`Render Offer`, () => {
  const match = {params: {id: `5`}};

  const offer = {
    adults: 3,
    avatar: `img/avatar-angelina.jpg`,
    bedrooms: 1,
    city: `Amsterdam`,
    cityId: 4,
    coord: [52.388540000000006, 4.899976],
    description: `A new spacious villa, one floor.`,
    id: 5,
    images: [`https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`],
    imgLink: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    isFavorite: true,
    isPremium: false,
    name: `Angelina`,
    price: 134,
    type: `test`,
    rating: 2.2,
    goods: [`goods`],
    threeImages: [`https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`],
    title: `The house among olive`
  };

  const loadCurrentOffer = jest.fn();
  const loadNearPlaces = jest.fn();
  const authorizationStatus = `yes`;
  const authInfo = {
    login: `Dusseldorf@mail.ru`,
    password: 1111
  };

  const updateFavorite = jest.fn();
  const good = `good`;
  const nearPlaces = [{
    name: `Amsterdam`,
    coord: [54, 45],
    id: 2
  }];
  const img = {id: 1, picture: `picture`};
  it(`Render Offer`, () => {
    const tree = renderer
      .create(
          <MemoryRouter>
            <Offer

              match={match}
              offer={offer}
              updateFavorite={updateFavorite}
              img={img}
              good={good}
              loadCurrentOffer={loadCurrentOffer}
              loadNearPlaces={loadNearPlaces}
              authorizationStatus={authorizationStatus}
              authInfo={authInfo}
              nearPlaces={nearPlaces}
              id={match}
            />
          </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});

