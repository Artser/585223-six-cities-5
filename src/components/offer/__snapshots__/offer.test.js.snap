// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Offer Render Offer 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <span
                className="header__nav-link header__nav-link--profile"
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <a
                  className="header__login"
                  href="/login"
                  onClick={[Function]}
                >
                  Sign in
                </a>
              </span>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__mark"
          >
            <span />
          </div>
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              The house among olive
            </h1>
            <button
              className="property__bookmark-button--active button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "40%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
               
              2.2
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              test
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              1
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              3
               adults
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              134
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                goods
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height="74"
                  src="img/avatar-angelina.jpg"
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A new spacious villa, one floor.
              </p>
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <ReviewList />
          </section>
        </div>
      </div>
      <div
        style={
          Object {
            "height": "579px",
          }
        }
      >
        <Map
          activeOffer={
            Array [
              52.388540000000006,
              4.899976,
            ]
          }
          coords={
            Array [
              Array [
                54,
                45,
              ],
            ]
          }
        />
      </div>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <div
          className="near-places__list places__list"
        >
          <article
            className="near-places__card place-card"
          >
            <div
              className="near-places__image-wrapper place-card__image-wrapper"
            >
              <a
                href="#"
              >
                <img
                  alt="Place image"
                  className="place-card__image"
                  height="200"
                  width="260"
                />
              </a>
            </div>
            <div
              className="place-card__info"
            >
              <div
                className="place-card__price-wrapper"
              >
                <div
                  className="place-card__price"
                >
                  <b
                    className="place-card__price-value"
                  >
                    €
                  </b>
                  <span
                    className="place-card__price-text"
                  >
                    / night
                  </span>
                </div>
                <button
                  className="place-card__bookmark-button place-card__bookmark-button--active button"
                  type="button"
                >
                  <svg
                    className="place-card__bookmark-icon"
                    height="19"
                    width="18"
                  >
                    <use
                      xlinkHref="#icon-bookmark"
                    />
                  </svg>
                  <span
                    className="visually-hidden"
                  >
                    In bookmarks
                  </span>
                </button>
              </div>
              <div
                className="place-card__rating rating"
              >
                <div
                  className="place-card__stars rating__stars"
                >
                  <span
                    style={
                      Object {
                        "width": false,
                      }
                    }
                  />
                  <span
                    className="visually-hidden"
                  >
                    Rating
                  </span>
                </div>
              </div>
              <h2
                className="place-card__name"
              >
                <a
                  href="#"
                />
              </h2>
              <p
                className="place-card__type"
              />
            </div>
          </article>
        </div>
      </section>
    </div>
  </main>
</div>
`;
